parquet-variant/src/main/java/org/apache/parquet/variant/Variant.java

```diff
@@ -0,0 +1,546 @@
+            return field.value;
+          }
+        }
+      } else {
+        int low = 0;
+        int high = size - 1;
+        while (low <= high) {
+          // Use unsigned right shift to compute the middle of `low` and `high`. This is not only a
+          // performance optimization, because it can properly handle the case where `low + high`
+          // overflows int.
+          int mid = (low + high) >>> 1;
+          ObjectField field =
+              getFieldAtIndex(mid, value, metadata, idSize, offsetSize, idStart, offsetStart, dataStart);
+          int cmp = field.key.compareTo(key);
+          if (cmp < 0) {
+            low = mid + 1;
+          } else if (cmp > 0) {
+            high = mid - 1;
+          } else {
+            return field.value;
+          }
+        }
+      }
+      return null;
+    });
+  }
+
+  /**
+   * A field in a Variant object.
+   */
+  public static final class ObjectField {
+    public final String key;
+    public final Variant value;
+
+    public ObjectField(String key, Variant value) {
+      this.key = key;
+      this.value = value;
+    }
+  }
+
+  /**
+   * Returns the ObjectField at the `index` slot. Return null if `index` is out of the bound of
+   * `[0, objectSize())`. `getType()` must be `Type.OBJECT`.
+   * @param index the index of the object field to get
+   * @return the Objectfield at the `index` slot, or null if `index` is out of bounds
+   */
+  public ObjectField getFieldAtIndex(int index) {
+    return VariantUtil.handleObject(value, pos, (size, idSize, offsetSize, idStart, offsetStart, dataStart) -> {
+      if (index < 0 || index >= size) {
+        return null;
+      }
+      return getFieldAtIndex(index, value, metadata, idSize, offsetSize, idStart, offsetStart, dataStart);
+    });
+  }
+
+  private static ObjectField getFieldAtIndex(
+      int index,
+      byte[] value,
+      byte[] metadata,
+      int idSize,
+      int offsetSize,
+      int idStart,
+      int offsetStart,
+      int dataStart) {
+    int id = VariantUtil.readUnsigned(value, idStart + idSize * index, idSize);
+    int offset = VariantUtil.readUnsigned(value, offsetStart + offsetSize * index, offsetSize);
+    String key = VariantUtil.getMetadataKey(metadata, id);
+    Variant v = new Variant(value, metadata, dataStart + offset);
+    return new ObjectField(key, v);
+  }
+
+  /**
+   * Returns the dictionary ID for the object field at the `index` slot.
+   * `getType()` must be `Type.OBJECT`.
+   * @param index the index of the object field to get the dictionary ID for
+   * @return the dictionary ID for the object field at the `index` slot
+   * @throws MalformedVariantException if `index` is out of bounds
+   */
+  public int getDictionaryIdAtIndex(int index) {
+    return VariantUtil.handleObject(value, pos, (size, idSize, offsetSize, idStart, offsetStart, dataStart) -> {
+      if (index < 0 || index >= size) {
+        throw VariantUtil.malformedVariant();
```

Comment:
Why is this a malformed variant, but the same issue with a field or array results in `null`?

I also don't think that this can assume that the variant is malformed in this case. The `index` is passed in by the caller and this is a public method. The index may not exist, but if so the problem is in the caller not the data.

Reply:
Actually, I don't see any uses of this. Can it be removed?